bliss/models/location_encoder.py

```diff
@@ -305,87 +303,81 @@ def get_n_source_prior_log_prob(self, detection_rate):
+
+        var_params_for_n_sources = {}
+        for k, param in self.variational_params.items():
+            indx_mat = getattr(self, k + "_indx")
+            indices = indx_mat[tile_n_sources.transpose(0, 1)].reshape(n_ptiles, -1)
+            var_param = torch.gather(vpf_padded, 1, indices)
+
+            var_params_for_n_sources[k] = rearrange(
+                var_param,
+                "np (ns d pd) -> ns np d pd",
+                np=n_ptiles,
+                ns=tile_n_sources.size(0),
+                d=self.max_detections,
+                pd=param["dim"],
+            )
+
+        # what?!? why is sigmoid(0) = 0?
+        loc_mean_func = lambda x: torch.sigmoid(x) * (x != 0).float()
+        # and why does location mean need to be transformed?
+        # can't we stick with the original logistic-normal parameterization here?
+        var_params_for_n_sources["loc_mean"] = loc_mean_func(var_params_for_n_sources["loc_mean"])
 
-        if squeeze:
-            var_params_for_n_sources = {
-                key: value.squeeze(0) for key, value in var_params_for_n_sources.items()
-            }
         return var_params_for_n_sources
 
     @property
     def variational_params(self):
-        # transform is a function applied directly on NN output.
         return {
-            "loc_mean": {"dim": 2, "transform": _loc_mean_func},
-            "loc_logvar": {"dim": 2, "transform": _identity_func},
-            "log_flux_mean": {"dim": self.n_bands, "transform": _identity_func},
-            "log_flux_logvar": {"dim": self.n_bands, "transform": _identity_func},
+            "loc_mean": {"dim": 2},
+            "loc_logvar": {"dim": 2},
+            "log_flux_mean": {"dim": self.n_bands},
+            "log_flux_logvar": {"dim": self.n_bands},
         }
 
-    # These methods are only used in testing or case studies, Do we need them or can
-    # they be moved to the code that they test? ------------------------
-
-    # --------------------------------------------------------------
-    def _get_var_params_for_n_sources(self, h, n_sources):
-        """Gets variational parameters for n_sources.
-
-        Arguments:
-            h: shape = (n_ptiles x dim_out_all)
-            n_sources: Tensor with shape (n_samples x n_ptiles)
-
-        Returns:
-            loc_mean.shape = (n_sample x n_ptiles x max_detections x len(x,y))
-        """
-        assert len(n_sources.shape) == 2
-
-        est_params = {}
-        for k, param in self.variational_params.items():
-            indx_mat = getattr(self, k + "_indx")
-            param_dim = param["dim"]
-            transform = param["transform"]
-            param = self._indx_h_for_n_sources(h, n_sources, indx_mat, param_dim)
-            param = transform(param)
-            est_params[k] = param
-
-        return est_params
-
     @staticmethod
-    def _get_normal_samples(mean, sd, tile_is_on_array):
+    def _sample_gated_normal(mean, sd, tile_is_on_array):
```

Comment:
It looks like this is could be the source of the sigmoid problem. Why are we zeroing out the "off" sources before and after the sigmoid transformation?

Reply:
oh good point, we should use `tile_is_on_array` to track "off sources" not depend on `loc` being 0.

Reply:
#529